File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.js
function solution(rows, columns, queries) {
    var answer = [];
    let prevMatrix = [];
    let count = 0;
    
    // 첫 행렬 만들기
    
    for (let i = 0; i < rows; i++) {
        let tempArray = [];
        for (let j = 0; j < columns; j++){
            count++;
            tempArray.push(count)
        }        
        prevMatrix.push(tempArray)
    }
    
    //
    
    queries.map((query) => {
        const x1 = query[0] - 1;
        const y1 = query[1] - 1;
        const x2 = query[2] - 1;
        const y2 = query[3] - 1;
        
        
        let copiedMatrix = JSON.parse(JSON.stringify(prevMatrix));
        let changed = []; // 바뀐 숫자들만 저장
        
        
        for (let x = x1; x < x2; x++) {
            copiedMatrix[x][y1] = prevMatrix[x+1][y1];
            changed.push(prevMatrix[x][y1]);
        }
        for (let y = y1; y < y2; y++) {
            copiedMatrix[x2][y] = prevMatrix[x2][y+1];
            changed.push(prevMatrix[x2][y]);
        }
        for (let x = x2; x > x1; x--) {
            copiedMatrix[x][y2] = prevMatrix[x-1][y2];
            changed.push(prevMatrix[x][y2]);
        }
        for (let y = y2; y > y1; y--) {
            copiedMatrix[x1][y] = prevMatrix[x1][y-1];
            changed.push(prevMatrix[x1][y]);
        }
        
        
        
        
        // 1. 다 돌았으면 바뀐 값들 중에서 가장 작은 수 찾기
        let minValue = Math.min(...changed)
        answer.push(minValue)
        
        // 2. copiedMatrix 를 prevMatrix로 바꾸기
        prevMatrix = JSON.parse(JSON.stringify(copiedMatrix))

            
    })
    
    
    return answer;
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.py
def rotate(r1, c1, r2, c2, matrix):
    first = matrix[r1][c1]
    min_value = first

    for i in range(r1, r2):
        matrix[i][c1] = matrix[i + 1][c1]
        min_value = min(min_value, matrix[i][c1])    
            
    for i in range(c1, c2):
        matrix[r2][i] = matrix[r2][i + 1]
        min_value = min(min_value, matrix[r2][i])    
        
    for i in range(r2, r1, -1):
        matrix[i][c2] = matrix[i - 1][c2]
        min_value = min(min_value, matrix[i][c2])
        
    for i in range(c2, c1, -1):
        matrix[r1][i] = matrix[r1][i - 1]
        min_value = min(min_value, matrix[r1][i])
    
    matrix[r1][c1 + 1] = first
    
    return min_value

def solution(rows, columns, queries):
    answer = []
    # 첫 배열 생성
    mtrx = [[(i + 1) + j * columns for i in range(columns)] for j in range(rows)]
    
    # rotate에서 돌려서 matrix를 바꾼다
    for [x1, y1, x2, y2] in queries:
        answer.append(rotate(x1 - 1, y1 - 1, x2 - 1, y2 - 1, mtrx))
    
    return answer